src/snap_lock.py

```diff
@@ -0,0 +1,119 @@
+"""Snap Lock Module using fcntl.flock and file-based reference counting."""
+
+import os
+import re
+import time
+from contextlib import contextmanager
+from typing import Generator
+import errno
+
+
+class SnapLockError(Exception):
+    """Base class for SnapLock error."""
+
+    pass
+
+
+class SnapLock:
+    """SnapLock provides mechanisms for singleton snaps using fcntl.flock and file-based reference counting.
+
+    Raises SnapLockError on any error.
+    """
+
+    def __init__(self, instance_name: str, lock_timeout_sec: int = 300):
+        # Replace all non-alphanumeric chars (except _-) with _
+        self.instance_name = re.sub(r'[^\w-]', '_', instance_name)
+        self.lock_timeout = lock_timeout_sec
+        self.separator = '--'
+
+        self.lock_dir = os.path.join('/run/lock', 'snap_locks')
```

Comment:
I came across [this](https://systemd.io/TEMPORARY_DIRECTORIES/) the other day, but not sure how to correctly implement 

> use your own sub-directory under /run/

Your impl seems right, just wanted to throw the ref, but you probably already know of it :)

Reply:
According to this article from systemd, it seems a subdirectory under `/run/` is preferred. According to Linux conventions, it seems we should use `/var/lock` (linked to `/run/lock`) for lock files. Combining these two, we end up with `/run/lock/some_subdirectory` is preferred. Also, I noticed that lock files have a naming convention that starts with "LCK..". I will also make this change.